### PR TITLE
feat/#74

### PR DESCRIPTION
**Pull request by DungNguyen2003**

### CODE REVIEW

File: src/app/components/ModalCustomization/index.tsx [hunk 1]
```diff
@@ -46,6 +48,7 @@ export const ModalCustomization = (props: Props) => {
     contentText,
     loading,
     textAgreeBtn,
+    userExamId,
```
Comment (nc-minh) [inline src/app/components/ModalCustomization/index.tsx:51]: Don't edit this, this component is generic for many cases, don't define a specific value in this. You can use route-dom's navigation instead to redirect with click . event.

File: src/app/components/ModalCustomization/index.tsx [hunk 2]
```diff
@@ -78,15 +81,20 @@ export const ModalCustomization = (props: Props) => {
               Cancel
             </Button>
             {textAgreeBtn && (
-              <LoadingButton
-                size="medium"
-                onClick={handleAgree}
-                loading={loading}
-                variant="contained"
-                color={colorBtn}
+              <Link
```
Comment (nc-minh) [inline src/app/components/ModalCustomization/index.tsx:84]: remove

File: src/app/containers/TestExamWrapper/index.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import RenderQuestion from "app/components/RenderQuestion";
+import { useGetUserExamByOwner } from "queries/userExam";
+
+const cx = classNames.bind(styles);
+
+export const TestExamWrapper = () => {
+  const { userExamId } = useParams();
+  const { data: userExamByOwner } = useGetUserExamByOwner(userExamId as string);
+
+  const examId = userExamByOwner?.original_exam;
```
Comment (nc-minh) [inline src/app/containers/TestExamWrapper/index.tsx:17]: useMemo for examId

File: src/app/containers/TestShow/index.tsx [hunk 4]
```diff
@@ -50,10 +49,13 @@ export const TestShow = () => {
         setDurationError("Thời gian làm bài cần lớn hơn 0 !!!");
         return;
       }
-      await mutateAsync({
+      const res = await mutateAsync({
         duration: Number(examDuration) * 60000,
         exam_id: exam?._id!,
       });
+
+      console.log("data", res);
```
Comment (nc-minh) [inline src/app/containers/TestShow/index.tsx:57]: remove

File: src/app/containers/TestShow/index.tsx [hunk 5]
```diff
@@ -141,6 +143,7 @@ export const TestShow = () => {
         loading={isLoading}
         textAgreeBtn="Start"
         colorBtn="success"
+        userExamId={data?._id} // error , lay duoc id cua user exam de su dung
```
Comment (nc-minh) [inline src/app/containers/TestShow/index.tsx:146]: remove

**Review by nc-minh — CHANGES_REQUESTED** [review]

File: src/app/components/ModalCustomization/index.tsx (hunk 1 above)
Comment (DungNguyen2003) [reply src/app/components/ModalCustomization/index.tsx:51]: but when click start button in popup duration, i used ModalCustomization component, if don't add the prop in here, how can i click button in popup to redirect page with link "exams/do-exam/${userExamId}"

Comment (nc-minh) [reply src/app/components/ModalCustomization/index.tsx:51]: Same as above you can use navigation https://reactrouter.com/en/main/hooks/use-navigate to redirect.

File: src/app/containers/TestExamWrapper/index.tsx [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+  const [openPopupSubmit, setOpenPopupSubmit] = useState(false);
+  const handleClosePopup = useCallback(() => {
+    setOpenPopupSubmit(false);
+  }, [openPopupSubmit]);
+
+  const handleOpenPopup = useCallback(() => {
+    setOpenPopupSubmit(true);
+  }, [openPopupSubmit]);
+
+  const changeAnswer = async (position: number, value: string) => {
```
Comment (nc-minh) [inline src/app/containers/TestExamWrapper/index.tsx:87]: useCallback here

File: src/app/containers/TestExamWrapper/index.tsx [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+      console.log(error);
+    }
+  };
+  useEffect(() => {
+    const updateAnswerId = arrUserAnswer
+      ? arrUserAnswer[lastUpdatedPosition || 0]
+      : "";
+
+    if (updateAnswerId !== "") {
+      mutateAsyncUpdateUserAnswer({
```
Comment (nc-minh) [inline src/app/containers/TestExamWrapper/index.tsx:108]: why put mutateAsyncUpdateUserAnswer in useEffect

File: src/app/containers/TestExamWrapper/index.tsx [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+      });
+    }
+  }, [
+    arrUserAnswer,
+    lastUpdatedPosition,
+    userExamByOwner?.user_answers[0]._id,
+    mutateAsyncUpdateUserAnswer,
+  ]);
+
+  const handleSubmitExam = async () => {
```
Comment (nc-minh) [inline src/app/containers/TestExamWrapper/index.tsx:124]: useCallback

File: src/app/containers/TestExamWrapper/index.tsx [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+      handleClosePopup();
+
+      const response: AxiosResponse = await http.get(
+        `/user-exams/${userExamId}`,
+      );
+      setScore(response.data.data.score);
+      setUserExamStatus(response.data.data.is_completed);
+    } catch (error) {}
+  };
+  const handleExit = () => {
```
Comment (nc-minh) [inline src/app/containers/TestExamWrapper/index.tsx:138]: useCallback

File: src/app/containers/TestShow/index.tsx [hunk 10]
```diff
@@ -26,6 +27,7 @@ export const TestShow = () => {
   const [durationError, setDurationError] = useState<string>("");
 
   const { mutateAsync, isLoading } = useCreateUserExam();
+  let userExamId: string = "";
```
Comment (nc-minh) [inline src/app/containers/TestShow/index.tsx:30]: don't need declare here

**Review by nc-minh — CHANGES_REQUESTED** [review]

**Comment by nc-minh** [comment]
pls resolve conflicts @DungNguyen2003

File: src/app/containers/TestExamWrapper/index.tsx (hunk 7 above)
Comment (DungNguyen2003) [reply src/app/containers/TestExamWrapper/index.tsx:108]: I thought wrong about result of api update user answer so I used useEffect hook,  i improved this code block by useCallback,

File: src/app/components/InProgress/index.tsx [hunk 11]
```diff
@@ -20,15 +20,18 @@ export const InProgress = () => {
...
-          <InProgressItem
-            key={userExam._id}
-            subjectName={userExam?.subject?.subject_name}
-            title={userExam.title}
-            userAnswer={userExam.user_answer_id}
-            totalQuestion={userExam.questions.length}
-          />
-        ))}
+        {userExams
+          ?.filter(userExam => !userExam.is_completed)
```
Comment (nc-minh) [inline src/app/components/InProgress/index.tsx:24]: Reduce the algorithm complexity.

File: src/app/containers/TestExamWrapper/TestExamWrapper.module.scss [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  display: flex;
+  justify-content: space-between;
+  border: 1px solid #ccc;
+  border-radius: 4px;
+  align-items: center;
+  padding: 12px;
+  position: sticky;
+  top: 0;
+  z-index: 1;
+  background-color: #c9c9c9;
```
Comment (nc-minh) [inline src/app/containers/TestExamWrapper/TestExamWrapper.module.scss:53]: don't use code color

File: src/app/containers/TestExamWrapper/index.tsx (hunk 3 above)
Comment (nc-minh) [reply src/app/containers/TestExamWrapper/index.tsx:28]: @DungNguyen2003

File: src/app/containers/TestExamWrapper/index.tsx [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+      );
+    }
+    if (userExamByOwner?.is_completed) {
+      setScore(userExamByOwner?.score!);
+    } else {
+      setScore(-1);
+    }
+    setUserExamStatus(userExamByOwner?.is_completed!);
+  }, [userExamByOwner]);
+  const [openPopupSubmit, setOpenPopupSubmit] = useState(false);
```
Comment (nc-minh) [inline src/app/containers/TestExamWrapper/index.tsx:75]: Grouping all useState to top.

File: src/app/components/InProgressItem/index.tsx [hunk 14]
```diff
@@ -53,6 +71,17 @@ export const InProgressItem = (props: Props) => {
...
+      <ModalCustomization
+        open={openPopup}
+        handleCancel={handleClosePopup}
+        handleAgree={handleSubmitContinue}
+        actionDefault
+        title="Bạn có chắc chắn muốn làm tiếp bài thi không?"
+        textAgreeBtn="Continue"
+        colorBtn="success"
+      >
+        <div></div>
```
Comment (nc-minh) [inline src/app/components/InProgressItem/index.tsx:83]: Why have a empty div inside ModalCustomizaion?

Comment (DungNguyen2003) [reply src/app/components/InProgressItem/index.tsx:83]: prop children in ModalCustomization component, to show contentText

**Pull request merged**